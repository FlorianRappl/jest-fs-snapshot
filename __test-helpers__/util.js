import path from 'path';

export const getFixture = name => path.resolve(process.cwd(), '__tests__/__fixtures__', name);

export const getAsar = name => path.resolve(process.cwd(), '__tests__/__fs_snapshots__', `${name}.asar`);
