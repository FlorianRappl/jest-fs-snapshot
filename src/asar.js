import asar from 'asar';
import asarDisk from 'asar/lib/disk';
import {getDirectoryTree} from './tree';

export function getAsarTree(archive) {
    return asarDisk.readArchiveHeaderSync(archive).header;
}

export function createAsarPackage(src, dest, {exclude, ...options} = {}) {
    const files = [];
    const metadata = {};
    function callback(path, stat) {
        let type;
        if (stat.isDirectory()) {
            type = 'directory';
        }
        if (stat.isFile()) {
            type = 'file';
        }
        if (stat.isSymbolicLink()) {
            type = 'symbolic';
        }

        if (!type) {
            throw new Error('ataatt');
        }

        files.push(path);
        metadata[path] = {type, stat};
    }

    getDirectoryTree(src, {exclude, callback});

    return asar.createPackageFromFiles(src, dest, files, metadata, options, err => (err ? reject(err) : resolve()));
}
