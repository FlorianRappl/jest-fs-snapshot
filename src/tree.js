import path from 'path';
import fs from 'fs';

export function getDirectoryTree(src, options = {}) {
    const stats = fs.lstatSync(src);

    if (options.callback) {
        options.callback(path.resolve(src), stats);
    }

    if (stats.isDirectory()) {
        const files = fs.readdirSync(src).reduce((res, child) => {
            if (!(options.exclude && options.exclude.test(child))) {
                res[child] = getDirectoryTree(path.resolve(src, child), options);
            }
            return res;
        }, {});

        return {files};
    }

    if (stats.isFile()) {
        return {
            size: stats.size,
        };
    }

    return {};
}

function getNodeType(node, pathToNode) {
    if (typeof node === 'object') {
        if (typeof node.files === 'object') return 'directory';
        if (typeof node.size === 'number') return 'file';
        if (typeof node.link === 'string') return 'symlink';
    }
    throw new Error(`Unknown node type '${JSON.stringify(node)}' at ${pathToNode}.`);
}

const ROOT = './';

export function compareDirectoryTrees({actual, expected, pathToNode = ROOT, onFile}) {
    const actualType = getNodeType(actual, pathToNode);
    const expectedType = getNodeType(expected, pathToNode);

    if (pathToNode === ROOT && (actualType !== 'directory' || expectedType !== 'directory')) {
        throw new Error(`The roots must be directories but got a ${actualType} and a ${expectedType}.`);
    }

    if (actualType !== expectedType) {
        throw new Error(`Nodes at '${pathToNode}' have different types: a ${actualType} and a ${expectedType}.`);
    }

    if (actualType === 'file') {
        if (onFile) {
            onFile({actual, expected, pathToNode});
            return null;
        }

        if (actual.size !== expected.size) {
            throw new Error(`Files at '${pathToNode}' have different size: ${actual.size} and ${expected.size}.`);
        }
    }

    if (actualType === 'directory') {
        const actualKeys = Object.keys(actual.files);
        const expectedKeys = Object.keys(expected.files);

        for (let i = 0, l = actualKeys.length; i < l; i++) {
            const key = actualKeys[i];
            const bKeyIndex = expectedKeys.indexOf(key);
            if (bKeyIndex === -1) {
                throw new Error(
                    `Directories at ${pathToNode} are different: expected directory doesn't have a node '${key}'.`
                );
            }
            expectedKeys.splice(bKeyIndex, 1);
        }

        if (expectedKeys.length > 0) {
            throw new Error(
                `Directories at ${pathToNode} are different: actual directory doesn't have node(s) '${expectedKeys.join(
                    "', '"
                )}'.`
            );
        }

        actualKeys.forEach(key => {
            compareDirectoryTrees({
                actual: actual.files[key],
                expected: expected.files[key],
                pathToNode: pathToNode + (pathToNode !== ROOT ? '/' : '') + key,
                onFile,
            });
        });
    }

    return null;
}
