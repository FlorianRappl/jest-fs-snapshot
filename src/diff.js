import path from 'path';
import fs from 'fs';
import isValidUTF8 from 'utf-8-validate';
import iconv from 'iconv-lite';
import diff from 'jest-diff';
import {NO_DIFF_MESSAGE} from 'jest-diff/build/constants';
import {EXPECTED_COLOR, RECEIVED_COLOR} from 'jest-matcher-utils';
import asarDisk from 'asar/lib/disk';
import {getDirectoryTree, compareDirectoryTrees} from './tree';
import {getAsarTree} from './asar';

export function getBufferEncoding(buffer) {
    const size = buffer.length;
    if (size === 0) {
        return 'utf8';
    }

    // UTF-8 BOM
    if (size >= 3 && buffer[0] === 0xef && buffer[1] === 0xbb && buffer[2] === 0xbf) {
        return 'utf8';
    }

    // UTF-16 BE BOM
    if (size >= 2 && buffer[0] === 0xfe && buffer[1] === 0xff) {
        return 'utf16-be';
    }

    // UTF-16 LE BOM
    if (size >= 2 && buffer[0] === 0xff && buffer[1] === 0xfe) {
        return 'utf16';
    }

    if (isValidUTF8(buffer)) {
        return 'utf8';
    }

    return 'binary';
}

export function compareBuffers(actualBuf, expectedBuf) {
    const actualEncoding = getBufferEncoding(actualBuf);
    const expectedEncoding = getBufferEncoding(expectedBuf);

    if (actualEncoding !== expectedEncoding) {
        return (
            ` Comparing files with different encodings. Expected ${EXPECTED_COLOR(expectedEncoding)}` +
            ` but received ${RECEIVED_COLOR(actualEncoding)}.`
        );
    }

    if (actualEncoding === 'binary') {
        return Buffer.compare(actualBuf, expectedBuf) === 0 ? NO_DIFF_MESSAGE : 'Comparing different binary files.';
    }

    const actualText = iconv.decode(actualBuf, actualEncoding);
    const expectedText = iconv.decode(expectedBuf, expectedEncoding);

    return diff(actualText, expectedText);
}

export function compareDirectoryToAsar(actualPath, expectedPath, options) {
    try {
        const actualTree = getDirectoryTree(actualPath, options);
        const expectedTree = getAsarTree(expectedPath);

        const asarFilesystem = asarDisk.readFilesystemSync(expectedPath);

        compareDirectoryTrees({
            actual: actualTree,
            expected: expectedTree,
            onFile({pathToNode}) {
                if (path.sep === '\\') {
                    // eslint-disable-next-line no-param-reassign
                    pathToNode = pathToNode.replace(/[/]/g, '\\');
                }

                const thisPath = path.resolve(actualPath, pathToNode);
                const actual = fs.readFileSync(thisPath);
                const expected = asarDisk.readFileSync(asarFilesystem, pathToNode, asarFilesystem.getFile(pathToNode));

                let compareFunc = compareBuffers;

                if (options.customCompare) {
                    const match = options.customCompare.filter(({check}) => check(thisPath))[0];
                    if (match) compareFunc = match.compare || compareFunc;
                }

                const res = compareFunc(actual, expected);

                if (res !== NO_DIFF_MESSAGE) {
                    const error = new Error(res);
                    error.pathToNode = pathToNode;
                    throw error;
                }
            },
        });
    } catch (error) {
        return error;
    }

    return null;
}

export {NO_DIFF_MESSAGE};
