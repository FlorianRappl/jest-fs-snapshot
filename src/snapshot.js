import fs from 'fs';
import path from 'path';
import filenamify from 'filenamify';
import {matcherHint, EXPECTED_COLOR, RECEIVED_COLOR} from 'jest-matcher-utils';
import {createAsarPackage} from './asar';
import {compareDirectoryToAsar} from './diff';

// eslint-disable-next-line import/prefer-default-export
export function matchFilesystemSnapshot({received, expected, context, name}) {
    const {testPath, currentTestName, isNot, snapshotState} = context;

    if (isNot) {
        throw new Error('Jest: `.not` cannot be used with `.toMatchFilesystemSnapshot()`.');
    }

    if (!snapshotState) {
        throw new Error('Jest: snapshot state must be initialized.');
    }

    const typeOfReceived = typeof received;
    if (typeOfReceived !== 'string') {
        throw new Error(`Jest: expected to receive a path to a directory but got ${typeOfReceived}.`);
    }

    const count = Number(snapshotState._counters.get(currentTestName) || 0);
    const testKey = `${path.basename(testPath)} ${currentTestName} ${count}`;
    const snapshotFilename = `${filenamify(testKey)}.snap.asar`;
    const snapshotsDir = path.resolve(testPath, '..', '__fs_snapshots__');

    const typeOfExpected = typeof expected;
    let expectedFullPath = expected;
    if (typeOfExpected === 'undefined') {
        expectedFullPath = path.resolve(snapshotsDir, snapshotFilename);
    } else if (typeOfExpected !== 'string') {
        throw new Error(`Jest: a matcher expected a path to a snapshot but got ${typeOfExpected}.`);
    }

    const receivedFullPath = path.resolve(received);
    if (!fs.existsSync(receivedFullPath)) {
        throw new Error(`Jest: expected path \`${received}\` doesn't exists.`);
    }

    const hasSnapshot = fs.existsSync(expectedFullPath);

    if (!hasSnapshot) {
        if (snapshotState._updateSnapshot === 'none') {
            return {
                name,
                pass: false,
                report: () =>
                    `New snapshot was ${RECEIVED_COLOR('not written')}. The update flag ` +
                    `must be explicitly passed to write a new snapshot.\n\n` +
                    `This is likely because this test is run in a continuous integration ` +
                    `(CI) environment in which snapshots are not written by default.`,
            };
        }

        createAsarPackage(receivedFullPath, expectedFullPath, {exclude: /node_modules/});

        snapshotState.added += 1;

        return {
            name,
            pass: true,
            message: () => '',
        };
    }

    const diffMessage = compareDirectoryToAsar(receivedFullPath, expectedFullPath, {
        exclude: /node_modules/,
        aAnnotation: 'Snapshot',
        bAnnotation: 'Received',
        expand: snapshotState.expand,
    });

    if (!diffMessage) {
        snapshotState.matched += 1;
        snapshotState._uncheckedKeys.delete(snapshotFilename);

        return {
            name,
            pass: true,
            message: () => '',
        };
    }

    if (snapshotState._updateSnapshot === 'all') {
        createAsarPackage(receivedFullPath, expectedFullPath, {exclude: /node_modules/});

        snapshotState.updated += 1;
        snapshotState._uncheckedKeys.delete(snapshotFilename);

        return {
            name,
            pass: true,
            message: () => '',
        };
    }

    snapshotState.unmatched += 1;
    snapshotState._uncheckedKeys.delete(snapshotFilename);

    const report = () =>
        `${RECEIVED_COLOR('Received value')} does not match ${EXPECTED_COLOR(
            `stored snapshot${diffMessage.pathToNode ? ` at ${diffMessage.pathToNode}` : ''}`
        )}.\n\n ${diffMessage.message}`;

    return {
        name,
        pass: false,
        message: () => `${matcherHint(`.${name}`, 'value', '')}\n\n${report()}`,
        report,
    };
}
