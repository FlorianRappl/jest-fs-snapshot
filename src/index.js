import {matchFilesystemSnapshot} from './snapshot';

// eslint-disable-next-line import/prefer-default-export
export function toMatchFilesystemSnapshot(received, expected) {
    return matchFilesystemSnapshot({received, expected, context: this, name: 'toMatchFilesystemSnapshot'});
}
