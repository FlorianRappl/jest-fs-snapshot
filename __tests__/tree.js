import {getDirectoryTree, compareDirectoryTrees} from '../src/tree';
import {getFixture} from '../__test-helpers__/util';

describe('compareDirectoryTrees()', () => {
    it('should throw if a node has an unknown type', () => {
        expect(() => compareDirectoryTrees({actual: 'foo', expected: {size: 0}})).toThrowErrorMatchingInlineSnapshot(
            `"Unknown node type '\\"foo\\"' at ./."`
        );

        expect(() =>
            compareDirectoryTrees({actual: {size: 0}, expected: {foo: 'bar'}})
        ).toThrowErrorMatchingInlineSnapshot(`"Unknown node type '{\\"foo\\":\\"bar\\"}' at ./."`);

        const actual = {
            files: {
                foo: 'bar',
            },
        };

        const expected = {
            files: {
                foo: {
                    size: 0,
                },
            },
        };

        expect(() => compareDirectoryTrees({actual, expected})).toThrowErrorMatchingInlineSnapshot(
            `"Unknown node type '\\"bar\\"' at ./foo."`
        );
    });

    it('should throw if the root is not a directory', () => {
        const options1 = {
            actual: {
                size: 1,
            },
            expected: {
                files: {},
            },
        };
        expect(() => compareDirectoryTrees(options1)).toThrowErrorMatchingInlineSnapshot(
            `"The roots must be directories but got a file and a directory."`
        );

        const options2 = {
            actual: {
                files: {},
            },
            expected: {
                size: 2,
            },
        };
        expect(() => compareDirectoryTrees(options2)).toThrowErrorMatchingInlineSnapshot(
            `"The roots must be directories but got a directory and a file."`
        );

        const options3 = {
            actual: {
                size: 1,
            },
            expected: {
                size: 2,
            },
        };
        expect(() => compareDirectoryTrees(options3)).toThrowErrorMatchingInlineSnapshot(
            `"The roots must be directories but got a file and a file."`
        );
    });

    it('should throw if node types are different', () => {
        const options1 = {
            actual: {
                files: {
                    foo: {
                        size: 1,
                    },
                },
            },
            expected: {
                files: {
                    foo: {
                        files: {},
                    },
                },
            },
        };
        expect(() => compareDirectoryTrees(options1)).toThrowErrorMatchingInlineSnapshot(
            `"Nodes at './foo' have different types: a file and a directory."`
        );

        const options2 = {
            actual: {
                files: {
                    foo: {
                        files: {},
                    },
                },
            },
            expected: {
                files: {
                    foo: {
                        size: 1,
                    },
                },
            },
        };
        expect(() => compareDirectoryTrees(options2)).toThrowErrorMatchingInlineSnapshot(
            `"Nodes at './foo' have different types: a directory and a file."`
        );

        const options3 = {
            actual: {
                files: {
                    foo: {
                        size: 1,
                    },
                    bar: {
                        files: {
                            baz: {
                                files: {},
                            },
                        },
                    },
                },
            },
            expected: {
                files: {
                    foo: {
                        size: 1,
                    },
                    bar: {
                        files: {
                            baz: {
                                size: 12,
                            },
                        },
                    },
                },
            },
        };
        expect(() => compareDirectoryTrees(options3)).toThrowErrorMatchingInlineSnapshot(
            `"Nodes at './bar/baz' have different types: a directory and a file."`
        );
    });

    it('should throw if files have different sizes', () => {
        const options1 = {
            actual: {
                files: {
                    foo: {
                        size: 1,
                    },
                    bar: {
                        files: {
                            baz: {
                                size: 11,
                            },
                        },
                    },
                },
            },
            expected: {
                files: {
                    foo: {
                        size: 2,
                    },
                    bar: {
                        files: {
                            baz: {
                                size: 11,
                            },
                        },
                    },
                },
            },
        };
        expect(() => compareDirectoryTrees(options1)).toThrowErrorMatchingInlineSnapshot(
            `"Files at './foo' have different size: 1 and 2."`
        );

        const options2 = {
            actual: {
                files: {
                    foo: {
                        size: 1,
                    },
                    bar: {
                        files: {
                            baz: {
                                size: 11,
                            },
                        },
                    },
                },
            },
            expected: {
                files: {
                    foo: {
                        size: 1,
                    },
                    bar: {
                        files: {
                            baz: {
                                size: 12,
                            },
                        },
                    },
                },
            },
        };
        expect(() => compareDirectoryTrees(options2)).toThrowErrorMatchingInlineSnapshot(
            `"Files at './bar/baz' have different size: 11 and 12."`
        );
    });

    it('should throw if directory contents are different', () => {
        const options1 = {
            actual: {
                files: {
                    foo: {
                        size: 1,
                    },
                    bar: {
                        size: 11,
                    },
                    baz: {
                        size: 21,
                    },
                },
            },
            expected: {
                files: {
                    foo: {
                        size: 1,
                    },
                    bar: {
                        size: 11,
                    },
                },
            },
        };
        expect(() => compareDirectoryTrees(options1)).toThrowErrorMatchingInlineSnapshot(
            `"Directories at ./ are different: expected directory doesn't have a node 'baz'."`
        );

        const options2 = {
            actual: {
                files: {
                    foo: {
                        size: 1,
                    },
                    bar: {
                        files: {
                            baz: {
                                size: 11,
                            },
                        },
                    },
                },
            },
            expected: {
                files: {
                    foo: {
                        size: 1,
                    },
                    bar: {
                        files: {
                            baz: {
                                size: 11,
                            },
                            qux: {
                                size: 21,
                            },
                        },
                    },
                },
            },
        };
        expect(() => compareDirectoryTrees(options2)).toThrowErrorMatchingInlineSnapshot(
            `"Directories at ./bar are different: actual directory doesn't have node(s) 'qux'."`
        );
    });

    it('should return null if trees are the same', () => {
        const options1 = {
            actual: {
                files: {
                    foo: {
                        files: {
                            bar: {
                                size: 1,
                            },
                            baz: {
                                files: {
                                    qux: {
                                        size: 2,
                                    },
                                },
                            },
                        },
                    },
                },
            },
            expected: {
                files: {
                    foo: {
                        files: {
                            bar: {size: 1},
                            baz: {
                                files: {
                                    qux: {
                                        size: 2,
                                    },
                                },
                            },
                        },
                    },
                },
            },
        };

        expect(compareDirectoryTrees(options1)).toBeNull();
    });
});

describe('getDirectoryTree()', () => {
    it('should read dir from `diff`', async () => {
        const fixturePath = getFixture('diff');
        const structure = getDirectoryTree(fixturePath);
        expect(structure).toMatchSnapshot();
    });

    it('should read dir from `encodings`', async () => {
        const fixturePath = getFixture('encodings');
        const structure = getDirectoryTree(fixturePath);
        expect(structure).toMatchSnapshot();
    });

    it('should read dir from `fixture-1`', async () => {
        const fixturePath = getFixture('fixture-1');
        const structure = getDirectoryTree(fixturePath);
        expect(structure).toMatchSnapshot();
    });
});
