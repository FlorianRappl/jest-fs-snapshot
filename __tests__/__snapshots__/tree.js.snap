// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDirectoryTree() should read dir from \`diff\` 1`] = `
Object {
  "files": Object {
    "emoji-1.txt": Object {
      "size": 25,
    },
    "emoji-2.txt": Object {
      "size": 29,
    },
    "lorem-ipsum-1.txt": Object {
      "size": 446,
    },
    "lorem-ipsum-2.txt": Object {
      "size": 343,
    },
  },
}
`;

exports[`getDirectoryTree() should read dir from \`encodings\` 1`] = `
Object {
  "files": Object {
    "ansi.txt": Object {
      "size": 314,
    },
    "big5.txt": Object {
      "size": 35,
    },
    "big5_B.txt": Object {
      "size": 940,
    },
    "cp936_with-bad-chars.txt": Object {
      "size": 39196,
    },
    "euc_jp.txt": Object {
      "size": 1284,
    },
    "euc_tw.txt": Object {
      "size": 35,
    },
    "euc_tw_B.txt": Object {
      "size": 306,
    },
    "gb.txt": Object {
      "size": 5,
    },
    "gb2.txt": Object {
      "size": 17,
    },
    "kr.txt": Object {
      "size": 42,
    },
    "latin.txt": Object {
      "size": 201,
    },
    "sjis.txt": Object {
      "size": 1284,
    },
    "utf16-be.txt": Object {
      "size": 100,
    },
    "utf16-be_bom.txt": Object {
      "size": 92,
    },
    "utf16-le_bom.txt": Object {
      "size": 92,
    },
    "utf32-le_bom.txt": Object {
      "size": 184,
    },
    "utf32_bom.txt": Object {
      "size": 184,
    },
    "utf8.txt": Object {
      "size": 102,
    },
    "utf8_bom.txt": Object {
      "size": 105,
    },
    "utf8_emojis.txt": Object {
      "size": 25,
    },
    "utf8_not-cp949.txt": Object {
      "size": 109,
    },
    "utf8_shishi.txt": Object {
      "size": 136,
    },
  },
}
`;

exports[`getDirectoryTree() should read dir from \`fixture-1\` 1`] = `
Object {
  "files": Object {
    "main.css": Object {
      "size": 31,
    },
    "tmp": Object {
      "files": Object {
        "tmp.js": Object {
          "size": 48,
        },
      },
    },
  },
}
`;
